File: packages/sdk/src/members.ts
import type {
  GameId,
  GroupId,
  Member,
  MemberId,
  MemberPermissionOverride,
  MemberStatus,
  Page,
  PageOptions,
  PermissionKey,
  RoleId,
  SetMemberNotesInput,
  UserId,
} from "@junjo.io/shared";
import { JunjoError } from "./errors.js";
import type { HttpClient } from "./http.js";
import { paginate } from "./pagination.js";
import { parseWireDate } from "./wire.js";

export interface WireMember {
  id: string;
  groupId: string;
  userId: string;
  status: string;
  roles: string[];
  metadata: Record<string, unknown>;
  notesPublic: string | null;
  notesPrivate: string | null;
  joinedAt: string;
  bannedUntil: string | null;
}

export function deserializeMember(w: WireMember): Member {
  return {
    id: w.id as MemberId,
    groupId: w.groupId as GroupId,
    userId: w.userId as UserId,
    status: w.status as MemberStatus,
    roles: w.roles.map((r) => r as RoleId),
    metadata: w.metadata,
    notesPublic: w.notesPublic,
    notesPrivate: w.notesPrivate,
    joinedAt: parseWireDate(w.joinedAt, "joinedAt"),
    bannedUntil: w.bannedUntil === null ? null : parseWireDate(w.bannedUntil, "bannedUntil"),
  };
}

export interface WireMemberPermissionOverride {
  groupId: string;
  userId: string;
  permission: string;
  grant: boolean;
  setAt: string;
  setBy: string | null;
}

export function deserializeMemberPermissionOverride(
  w: WireMemberPermissionOverride,
): MemberPermissionOverride {
  return {
    groupId: w.groupId as GroupId,
    userId: w.userId as UserId,
    permission: w.permission as PermissionKey,
    grant: w.grant,
    setAt: parseWireDate(w.setAt, "setAt"),
    setBy: w.setBy === null ? null : (w.setBy as UserId),
  };
}

/** Options for {@link MembersApi.list}. */
export interface ListMembersOptions extends PageOptions {
  /**
   * Filter to one or more statuses. Omit for all statuses (the default).
   * Common shapes: ["active"] for "show me current members",
   * ["banned"] for the moderation panel.
   */
  status?: MemberStatus[];
  signal?: AbortSignal;
  timeoutMs?: number;
}

/**
 * Group members: lookups, listing, per-member metadata and notes, role
 * assignment, and per-member permission overrides.
 */
export class MembersApi {
  constructor(private readonly http: HttpClient) {}

  async get(
    groupId: GroupId,
    userId: UserId,
    opts?: { signal?: AbortSignal; timeoutMs?: number },
  ): Promise<Member | null> {
    try {
      const wire = await this.http.get<WireMember>(
        `/v1/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(userId)}`,
        { signal: opts?.signal, timeoutMs: opts?.timeoutMs },
      );
      return deserializeMember(wire);
    } catch (err) {
      if (err instanceof JunjoError && err.code === "not_found") return null;
      throw err;
    }
  }

  async getById(
    id: MemberId,
    opts?: { signal?: AbortSignal; timeoutMs?: number },
  ): Promise<Member | null> {
    try {
      const wire = await this.http.get<WireMember>(`/v1/members/${encodeURIComponent(id)}`, {
        signal: opts?.signal,
        timeoutMs: opts?.timeoutMs,
      });
      return deserializeMember(wire);
    } catch (err) {
      if (err instanceof JunjoError && err.code === "not_found") return null;
      throw err;
    }
  }

  /**
   * Async-iterator wrapper over `list(...)`. Walks every page until
   * `nextCursor` is null. Combine with `status` to iterate all banned
   * members in a group, all kicked members, etc.
   */
  listAll(groupId: GroupId, opts?: Omit<ListMembersOptions, "cursor">): AsyncGenerator<Member> {
    return paginate((cursor) => this.list(groupId, { ...opts, cursor }));
  }

  async list(groupId: GroupId, opts?: ListMembersOptions): Promise<Page<Member>> {
    const params = new URLSearchParams();
    if (opts?.limit !== undefined) params.set("limit", String(opts.limit));
    if (opts?.cursor !== undefined) params.set("cursor", opts.cursor);
    if (opts?.status !== undefined && opts.status.length > 0) {
      params.set("status", opts.status.join(","));
    }
    const qs = params.toString();
    const base = `/v1/groups/${encodeURIComponent(groupId)}/members`;
    const path = qs ? `${base}?${qs}` : base;
    const wire = await this.http.get<{ items: WireMember[]; nextCursor: string | null }>(path, {
      signal: opts?.signal,
      timeoutMs: opts?.timeoutMs,
    });
    return {
      items: wire.items.map(deserializeMember),
      nextCursor: wire.nextCursor,
    };
  }

  async listForUser(
    userId: UserId,
    opts?: { gameId?: GameId; signal?: AbortSignal; timeoutMs?: number },
  ): Promise<Member[]> {
    const params = new URLSearchParams();
    if (opts?.gameId !== undefined) params.set("gameId", opts.gameId);
    const qs = params.toString();
    const base = `/v1/users/${encodeURIComponent(userId)}/members`;
    const path = qs ? `${base}?${qs}` : base;
    const wire = await this.http.get<WireMember[]>(path, {
      signal: opts?.signal,
      timeoutMs: opts?.timeoutMs,
    });
    return wire.map(deserializeMember);
  }

  async setMetadata(
    groupId: GroupId,
    userId: UserId,
    metadata: Record<string, unknown>,
    opts?: { signal?: AbortSignal; timeoutMs?: number },
  ): Promise<Member> {
    const wire = await this.http.patch<WireMember>(
      `/v1/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(userId)}`,
      { metadata },
      { signal: opts?.signal, timeoutMs: opts?.timeoutMs },
    );
    return deserializeMember(wire);
  }

  async setNotes(
    groupId: GroupId,
    userId: UserId,
    input: SetMemberNotesInput,
    opts?: { signal?: AbortSignal; timeoutMs?: number },
  ): Promise<Member> {
    const body: Record<string, string | null> = {};
    if (input.notesPublic !== undefined) body.notesPublic = input.notesPublic;
    if (input.notesPrivate !== undefined) body.notesPrivate = input.notesPrivate;
    const wire = await this.http.patch<WireMember>(
      `/v1/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(userId)}`,
      body,
      { signal: opts?.signal, timeoutMs: opts?.timeoutMs },
    );
    return deserializeMember(wire);
  }

  async assignRole(
    groupId: GroupId,
    userId: UserId,
    roleId: RoleId,
    opts?: { actorUserId?: UserId; signal?: AbortSignal; timeoutMs?: number },
  ): Promise<Member> {
    const body = opts?.actorUserId !== undefined ? { actorUserId: opts.actorUserId } : undefined;
    const wire = await this.http.post<WireMember>(
      `/v1/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(userId)}/roles/${encodeURIComponent(roleId)}`,
      body,
      { signal: opts?.signal, timeoutMs: opts?.timeoutMs },
    );
    return deserializeMember(wire);
  }

  async removeRole(
    groupId: GroupId,
    userId: UserId,
    roleId: RoleId,
    opts?: { actorUserId?: UserId; signal?: AbortSignal; timeoutMs?: number },
  ): Promise<Member> {
    const body = opts?.actorUserId !== undefined ? { actorUserId: opts.actorUserId } : undefined;
    const wire = await this.http.delete<WireMember>(
      `/v1/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(userId)}/roles/${encodeURIComponent(roleId)}`,
      body,
      { signal: opts?.signal, timeoutMs: opts?.timeoutMs },
    );
    return deserializeMember(wire);
  }

  async overridePermission(
    groupId: GroupId,
    userId: UserId,
    permission: PermissionKey,
    grant: boolean,
    opts?: { signal?: AbortSignal; timeoutMs?: number },
  ): Promise<MemberPermissionOverride> {
    const wire = await this.http.post<WireMemberPermissionOverride>(
      `/v1/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(userId)}/permissions/${encodeURIComponent(permission)}`,
      { grant },
      { signal: opts?.signal, timeoutMs: opts?.timeoutMs },
    );
    return deserializeMemberPermissionOverride(wire);
  }

  async clearPermissionOverride(
    groupId: GroupId,
    userId: UserId,
    permission: PermissionKey,
    opts?: { signal?: AbortSignal; timeoutMs?: number },
  ): Promise<void> {
    await this.http.delete<unknown>(
      `/v1/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(userId)}/permissions/${encodeURIComponent(permission)}`,
      undefined,
      { signal: opts?.signal, timeoutMs: opts?.timeoutMs },
    );
  }

  async listPermissionOverrides(
    groupId: GroupId,
    userId: UserId,
    opts?: { signal?: AbortSignal; timeoutMs?: number },
  ): Promise<MemberPermissionOverride[]> {
    const wire = await this.http.get<WireMemberPermissionOverride[]>(
      `/v1/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(userId)}/permissions`,
      { signal: opts?.signal, timeoutMs: opts?.timeoutMs },
    );
    return wire.map(deserializeMemberPermissionOverride);
  }
}

File: packages/sdk-roblox/src/members.lua
--!nonstrict
-- Nonstrict, not strict: cross-module `require(script.Parent.X)` types
-- and the metatable-OOP idiom below need the Roblox definition files to
-- pass strict analysis, which CI cannot run yet. Public signatures
-- carry annotations regardless.
--
-- Members namespace. Mirrors the TypeScript SDK's `junjo.members`
-- surface (member lookups + role / permission edits + metadata / notes
-- mutations). Membership lifecycle (invite / accept / leave / kick) is
-- exposed via `junjo.groups` to match the TS SDK shape.
--
-- Roblox game servers are TRUSTED callers: they hold the per-game
-- `jk_` key, so the full members surface (roles, overrides, notes) is
-- legitimately available server-side. Never ship the key to clients.

local JunjoError = require(script.Parent.JunjoError)
local tryGet = require(script.Parent.TryGet)
local pageAll = require(script.Parent.PageAll)
local Types = require(script.Parent.Types)

type Member = Types.Member
type Page<T> = Types.Page<T>

local Members = {}
Members.__index = Members

-- Shallow-copies opts so a pageAll loop can vary the cursor without
-- mutating the caller's table.
local function withCursor(opts, cursor: string?)
	local merged = {}
	if opts then
		for k, v in pairs(opts) do
			merged[k] = v
		end
	end
	merged.cursor = cursor
	return merged
end

function Members.new(http)
	local self = setmetatable({}, Members)
	self._http = http
	return self
end

local function memberPath(self, groupId: string, userId: string): string
	return "/v1/groups/" .. self._http:encode(groupId)
		.. "/members/" .. self._http:encode(userId)
end

-- ============================================================
-- Lookups
-- ============================================================

function Members:get(groupId: string, userId: string)
	return tryGet(self._http, memberPath(self, groupId, userId))
end

function Members:getById(id: string)
	return tryGet(self._http, "/v1/members/" .. self._http:encode(id))
end

-- `opts.status` filters to one or more member statuses (sent to the
-- server as a comma-separated list). Omit it (or pass an empty array)
-- for all statuses. Common shapes: { "active" } for "show me current
-- members", { "banned" } for a moderation panel.
function Members:list(groupId: string, opts: { limit: number?, cursor: string?, status: { string }? }?): Page<Member>
	local query = {}
	if opts and opts.limit ~= nil then
		table.insert(query, "limit=" .. self._http:encode(opts.limit))
	end
	if opts and opts.cursor ~= nil then
		table.insert(query, "cursor=" .. self._http:encode(opts.cursor))
	end
	if opts and opts.status ~= nil and #opts.status > 0 then
		table.insert(query, "status=" .. self._http:encode(table.concat(opts.status, ",")))
	end
	local path = "/v1/groups/" .. self._http:encode(groupId) .. "/members"
	if #query > 0 then
		path = path .. "?" .. table.concat(query, "&")
	end
	return self._http:get(path)
end

-- Generic-for iterator over `list(...)` that walks every page until
-- `nextCursor` is nil. Combine with `opts.status` to iterate all
-- banned members in a group, all kicked members, etc. `opts.cursor`
-- is owned by the iterator and ignored if supplied.
function Members:listAll(groupId: string, opts: { limit: number?, status: { string }? }?): () -> Member?
	return pageAll(function(cursor)
		return self:list(groupId, withCursor(opts, cursor))
	end)
end

function Members:listForUser(userId: string, opts: { gameId: string? }?)
	local path = "/v1/users/" .. self._http:encode(userId) .. "/members"
	if opts and opts.gameId ~= nil then
		path = path .. "?gameId=" .. self._http:encode(opts.gameId)
	end
	return self._http:get(path)
end

-- ============================================================
-- Metadata + notes
-- ============================================================

function Members:setMetadata(groupId: string, userId: string, metadata: { [string]: any })
	if metadata == nil then
		JunjoError.raise(
			"members:setMetadata(groupId, userId, metadata) requires a metadata table",
			"invalid_config",
			nil
		)
	end
	return self._http:patch(memberPath(self, groupId, userId), { metadata = metadata })
end

-- Partial notes update. Pass `Junjo.Null` for either field to clear it
-- on the server; omit a field (or set it to nil) to leave it unchanged.
function Members:setNotes(groupId: string, userId: string, input: { notesPublic: any?, notesPrivate: any? })
	if input == nil then
		JunjoError.raise(
			"members:setNotes(groupId, userId, input) requires an input table",
			"invalid_config",
			nil
		)
	end
	local body = {}
	if input.notesPublic ~= nil then body.notesPublic = input.notesPublic end
	if input.notesPrivate ~= nil then body.notesPrivate = input.notesPrivate end
	return self._http:patch(memberPath(self, groupId, userId), body)
end

-- ============================================================
-- Role assignment
-- ============================================================

function Members:assignRole(groupId: string, userId: string, roleId: string)
	return self._http:post(
		memberPath(self, groupId, userId) .. "/roles/" .. self._http:encode(roleId),
		nil
	)
end

function Members:removeRole(groupId: string, userId: string, roleId: string)
	return self._http:delete(
		memberPath(self, groupId, userId) .. "/roles/" .. self._http:encode(roleId)
	)
end

-- ============================================================
-- Permission overrides
-- ============================================================

function Members:overridePermission(groupId: string, userId: string, permission: string, grant: boolean)
	return self._http:post(
		memberPath(self, groupId, userId) .. "/permissions/" .. self._http:encode(permission),
		{ grant = grant }
	)
end

function Members:clearPermissionOverride(groupId: string, userId: string, permission: string)
	self._http:delete(
		memberPath(self, groupId, userId) .. "/permissions/" .. self._http:encode(permission)
	)
end

function Members:listPermissionOverrides(groupId: string, userId: string)
	return self._http:get(memberPath(self, groupId, userId) .. "/permissions")
end

return Members
